pkg/proxy/topology.go

```diff
@@ -0,0 +1,79 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package proxy
+
+import (
+	v1 "k8s.io/api/core/v1"
+)
+
+// FilterTopologyEndpoint returns the appropriate endpoints based on the cluster
+// topology.
+// It uses current node's labels which contains topology information, and use
+// topologyKeys to filter out the appropriate endpoints.
+// If both target endpoint's node and current node has a label which key equals to
+// topologyKeys[0], and value equals to each other, the endpoint will be chosen;
+// Only if no endpoints been chosen for topologyKeys[0] should endpoints for
+// toplogyKeys[1] be considered, and so on.
+// If topologyKeys is specified, but all entries have no endpoints that match the
+// the topology of the current node, the service has no endpoints for the current
+// node, and connections should fails.
+// The special value "" in topologyKeys may be used to mean "any node", this
+// catch-all value, if used, only makes sense as the last value in the list.
+// If topologyKeys is not specified or empty, no topology constraints will be
+// applied, it will return all endpoints.
+func FilterTopologyEndpoint(nodeLabels map[string]string, topologyKeys []string, endpoints []Endpoint) []Endpoint {
+	// Do not filter endpoints in the following cases:
+	// 1. Service without topology keys.
+	// 2. Can not get current node labels (e.g. hostname is different with node name).
+	if nodeLabels == nil || len(topologyKeys) == 0 {
+		return endpoints
```

Comment:
The following cases will not filter endpoints (return all endpoints):
1. Service without topology keys.
2. Can not get current node labels which contains topology information (e.g. hostname is different with node name so we can not watch current node, or node was out of cluster, but still want to use kube-proxy to sync rule to visit cluster ip for some reason).